### PR TITLE
[dynamo] Fix support for nn.Parameter constructor (part 1)

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #121610
* #120965
* #121609
* __->__ #120163
* #121086

This captures calls to `torch.nn.Parameter` by lifting them to graph inputs.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: test/inductor/test_cpu_repro.py [hunk 1]
```diff
@@ -2878,6 +2879,35 @@ def forward(self, x):
...
+                super().__init__()
+                self.conv = nn.Conv2d(in_channels=3, out_channels=5, kernel_size=3)
+                self.batchnorm = nn.BatchNorm2d(num_features=5)
+                self.conv_weight = torch.randn(5, 3, 3, 3)
+                self.conv_bias = torch.randn(5)
+
+            def forward(self, x):
+                self.conv.weight = nn.Parameter(self.conv_weight)
+                self.conv.bias = nn.Parameter(self.conv_bias, requires_grad=False)
+                self.conv.eval()
```
Comment (albanD) [inline test/inductor/test_cpu_repro.py:2895]: This is not cool haha
It should be a no-op though so not sure it changes much? Since you don't test it explicitly, I would remove it.

Unless you meant to call this on the batchnorm where it will change its behavior!

File: test/inductor/test_cpu_repro.py [hunk 2]
```diff
@@ -2878,6 +2879,35 @@ def forward(self, x):
...
+                x = self.conv(x)
+                x = self.batchnorm(x)
+                x = F.relu(x)
+                return x
+
+        input_tensor = torch.randn(1, 3, 10, 10)
+        func = Model2().to("cpu")
+
+        with torch.no_grad():
+            func.train(False)
```
Comment (albanD) [inline test/inductor/test_cpu_repro.py:2905]: This makes the eval() call above a noop?

File: torch/_dynamo/output_graph.py [hunk 3]
```diff
@@ -1575,6 +1582,13 @@ def add_graph_finalizer(
     ) -> None:
         self.register_finalizer_fns.append(register_finalizer)
 
+    def example_value_from_input_node(self, node: torch.fx.Node):
+        """Extract the non-fake example tensor"""
+        if node.op == "placeholder":
+            return node.meta["grapharg"].example
+        assert node.op == "get_attr"
+        return self.nn_modules[node.target]  # type: ignore[index]
```
Comment (albanD) [inline torch/_dynamo/output_graph.py:1590]: Is "get_attr" a nn Module specific op?

File: torch/_dynamo/variables/torch.py [hunk 4]
```diff
@@ -759,3 +764,50 @@ def handle_ntuple(value):
...
+        from .builder import VariableBuilder
+
+        varname = tx.output.new_var()
+
+        # construct the nn.Parmeter before the graph save it to varname
+        cg = PyCodegen(tx)
+        cg.load_import_from("torch.nn", "Parameter")
+        cg(data.source)
+        cg(variables.ConstantVariable(requires_grad))
+        cg.call_function(2, True)
```
Comment (albanD) [inline torch/_dynamo/variables/torch.py:798]: I can't find the definition of this function in torch/_dynamo/codegen.py where PyCodegen is defined. Where is it?

**Review by albanD — COMMENTED** [review]
I'm not the best person to review the dynamo code so this is mostly questions haha

File: test/inductor/test_cpu_repro.py (hunk 1 above)
Comment (jansel) [reply test/inductor/test_cpu_repro.py:2895]: This is just copied from the repro provided in https://github.com/pytorch/pytorch/issues/99569

I agree it doesn't make much sense, but want to make sure we don't break on that user report.

File: test/inductor/test_cpu_repro.py (hunk 2 above)
Comment (jansel) [reply test/inductor/test_cpu_repro.py:2905]: Yes

File: torch/_dynamo/output_graph.py (hunk 3 above)
Comment (jansel) [reply torch/_dynamo/output_graph.py:1590]: Yeah, dynamo only uses `get_attr` for nn modules (and tensors stored on nn modules).

File: torch/_dynamo/variables/torch.py (hunk 4 above)
Comment (jansel) [reply torch/_dynamo/variables/torch.py:798]: https://github.com/pytorch/pytorch/blob/bcf35c6ae62bb6560befa3550e37a8283944e5f4/torch/_dynamo/codegen.py#L323-L324

Comment (albanD) [reply torch/_dynamo/variables/torch.py:798]: Ho it was added last week and is not in the folder I was checking. Thanks!

**Review by albanD — APPROVED** [review]
The change sounds good to me. You might want a second opinion from someone familiar with Dynamo though.

**Review by yanboliang — APPROVED** [review]
